microovn/src/charm.py

```diff
@@ -92,49 +147,74 @@ def find_token(self, relation):
 
     def join_with_token(self, token):
         self.unit.status = ops.MaintenanceStatus("Joining cluster")
-        os.system("microovn cluster join {}".format(token))
+        error, _ = call_microovn_command("cluster", "join", token)
+        if error:
+            return False
         self._stored.in_cluster=True
         self.unit.status = ops.ActiveStatus("Joined cluster")
+        return True
 
     def _on_install(self, event: ops.InstallEvent):
         self.unit.status = ops.MaintenanceStatus("installing microovn snap")
-        os.system("snap install microovn --channel latest/edge")
+
+        subprocess.run(["snap", "install", "microovn", "--channel", "latest/edge"])
         self.unit.status = ops.MaintenanceStatus("waiting for microovn ready")
-        os.system("microovn waitready")
+        retries = 0
+        while not call_microovn_command("waitready"):
+            retries+=1
+            self.unit.status = ops.MaintenanceStatus(
+                "microovn waitready failed, retry {0}".format(retries))
         self.unit.status = ops.MaintenanceStatus("waiting for cluster setup")
 
-    def _on_start(self, event: ops.StartEvent):
-        if self.unit.is_leader():
-            # only bootstrap if leader
-            # TODO: fix cluster bootstrapped, leader changes to new node not in
-            #       cluster, bootstrapps new cluster can be fixed via having some
-            #       other characteristic determine who manages this side of the
-            #       mirror ie: microcluster leader
-            os.system(f"microovn cluster bootstrap")
-            self._stored.in_cluster = True
-            self.unit.status = ops.ActiveStatus("cluster bootstrapped")
+    def any_token_exists(self, relation):
+        for unit in relation.units:
+            for k in relation.data[unit].keys():
+                if k.startswith(MIRROR_PREFIX) and \
+                   relation.data[unit][k] != "empty":
+                    return True
+        return False
 
     def _handle_relation_created(self, event: ops.RelationCreatedEvent):
         self.add_hostname(WORKER_RELATION)
-        self.update_mirror_state(event.relation.data)
-        if self.unit.is_leader():
-            self.update_tokens(event.relation)
+
+        token_in_cluster = False
+        if relation := self.model.get_relation(WORKER_RELATION):
+            token_in_cluster = self.any_token_exists(relation)
+
+        if not self._stored.in_cluster and self.unit.is_leader() and \
+           not token_in_cluster:
+            error, _ = call_microovn_command("cluster", "bootstrap")
+            if error:
+                logger.error("{0} unable to bootstrap cluster".format(
+                    get_hostname()))
+                self.unit.status = ops.BlockedStatus("unable to bootstrap cluster")
```

Comment:
I like the way you bring fatal errors to the attention of the agent/person operating the operator!